File: discord-frontend.js
// Discord slash command frontend for http_backend.py

const axios = require("axios");
const { Client, Intents } = require("discord.js");

const { discord_slash_prefix, discord_token, backend_port_number } = require("./config.json");
const sailorServiceURL = `http://localhost:${backend_port_number}`;

const client = new Client({ intents: [ Intents.FLAGS.GUILDS ] });

client.on("ready", onceReady);
client.on("interactionCreate", onInteractionCreate);

client.login(discord_token);

// Standard ready message. Also sets playing status
function onceReady() {
    console.info(":3");
    client.user.setPresence({
        status: "online",
        activities: [{
            name: `Type /${discord_slash_prefix} help for help!`
        }]
    });
}

// Command handling
async function onInteractionCreate(interaction) {
	if (!interaction.isCommand()) return;

    await interaction.deferReply();

    let commandArguments = await interaction.options.getString("input", false);
    if (commandArguments === undefined) return await deleteOriginalReply(interaction);

    let channel = interaction.channel;
	let commandName = interaction.commandName;
    let fullCommand;
    if (commandName === discord_slash_prefix) {
        if (commandArguments) fullCommand = commandArguments;
        else return await deleteOriginalReply(interaction);
    }
    else {
        if (commandArguments) fullCommand = `${commandName} ${commandArguments}`;
        else fullCommand = commandName;
    }

    console.info(
        `id=${interaction.id} ` +
        `user=${interaction.user.tag} ` +
        `userId=${interaction.user.id} ` +
        `| ${fullCommand}`
    );

    let requestBody = {
        id: `discord.js:${interaction.id}`,
        message: fullCommand,
        is_owner: false,
        character_limit: 2000,
        format_name: "discord"
    }
    if (channel && typeof channel.name === "string") {
        requestBody.channel_name = channel.name;
    }

    axios
        .post(sailorServiceURL, requestBody)
        .then(async response => await doActions(response, interaction, channel, false))
        .catch(async error => {
            try { await doActions(error.response, interaction, channel, true); }
            catch {
                let errorMessage;
                if (error.code) errorMessage = `An error occurred: ${error.code}`
                else errorMessage = "An unknown error occurred.";

                console.error(`id=${interaction.id} | ${errorMessage}`);
                if (error.code === "ECONNREFUSED") {
                    await interaction.editReply("My brain stopped working. Please contact my owner. :<");
                }
                else if (error.code !== "ECONNRESET") {
                    await interaction.editReply(errorMessage);
                }
            }
        });
}

// Delete original reply to interaction (assumes reply or deferReply is called already)
async function deleteOriginalReply(interaction) {
    await interaction.editReply("…");
    await interaction.deleteReply();
}

// Truncates a string to a single line for logging
function toOneLiner(string, maxLength=75) {
    let oneLiner = string.split("\n").join(" ");
    if (oneLiner.length > maxLength) {
        oneLiner = oneLiner.substring(0, maxLength-1) + "…";
    }
    if (oneLiner.length === 0) oneLiner = "<empty>";
    return oneLiner;
}

// Execute a single action requested by the backend
function doAction(action, interaction, channel, isError=false) {
    if (typeof action.type !== "string" || typeof action.value !== "string") {
        return;
    }
    let logMessage = `id=${interaction.id} actionType=${action.type}`;
    switch (action.type) {
        case "rename_channel":
            if (!channel) {
                console.warn(`${logMessage} | Channel rename can't be done!`);
                return;
            }
            logMessage += (
                ` channelId=${channel.id}` +
                ` channelOldName=${channel.name}` +
                ` channelNewName=${toOneLiner(action.value)}`
            );
            channel.edit({ name: action.value })
                .then(() => console.info(`${logMessage} | Rename succeeded`))
                .catch(error => console.warn(`${logMessage} | Rename failed: ${error}`));
            break;
        case "reply":
            logMessage += ` | ${toOneLiner(action.value)}`;
            if (isError) {
                console.error(logMessage);
                interaction.followUp({ content: action.value, ephemeral: true });
            }
            else {
                console.info(logMessage);
                interaction.followUp(action.value);
            }
            break;
        default:
            console.warn(`${logMessage} | Unsupported action`);
    }
}

// Read over a response with a list of actions and perform the actions in sequence
async function doActions(response, interaction, channel, isError=false) {
    if (isError) await deleteOriginalReply(interaction);
    if (response.data.length && response.status !== 404) {
        let numReplies = 0;
        response.data.forEach(action => {
            doAction(action, interaction, channel, isError);
            if (action.type === "reply") numReplies++;
        });
        if (numReplies === 0) await deleteOriginalReply(interaction);
    }
    else {
        // NOTE: Technically not true, but usually true, and is more user-friendly
        await interaction.followUp({
            content: `Not a valid command. Type /${discord_slash_prefix} help for a list of commands.`,
            ephemeral: true
        });
    }
}

File: discord_frontend.py
"""
A Discord frontend for http_backend.py

Requires Python 3.6+ and discord.py 1.0 or higher.
"""

# pylint: disable=invalid-name,broad-except

import json
from typing import List

import aiohttp
import discord

from sailor_fox.helpers import create_logger, to_one_liner

logger = create_logger("discord")


def get_prefix(text: str, prefixes_: List[str]):
    """
    If a text string starts with a substring, return the substring and the text minus the first instance of the
    substring; otherwise return None and the text.
    """
    for prefix in prefixes_:
        if text.startswith(prefix):
            return prefix, text[len(prefix):].lstrip()
    return None, text


def main():
    """Factory to create and start the bot."""

    with open("config.json") as config_file:
        config = json.load(config_file)
    backend_port_number = config.get("backend_port_number", 9980)
    prefixes = []

    assert (isinstance(config.get("discord_token"), str)), "Bot token not valid."

    client = discord.AutoShardedClient()
    client_session = aiohttp.ClientSession(loop=client.loop)

    @client.event
    async def on_ready():
        """Set the bot's playing status to the help command."""

        main_prefix = config.get("prefix", "")
        prefixes.append(main_prefix)
        prefixes.append(client.user.mention)
        prefixes.append(client.user.mention.replace("@", "@!"))  # Mentions are cursed

        prefix_space = " " if main_prefix and main_prefix[-1].isalnum() else ""
        game = discord.Game(name=f"Type {main_prefix}{prefix_space}help for help!")
        await client.change_presence(activity=game)

    @client.event
    async def on_message(message: discord.Message):
        """Handle on_message events from Discord and forward them to the processor."""

        if message.author.bot:
            return

        prefix_or_none, message_text = get_prefix(message.content, prefixes)
        if prefix_or_none is None or not message_text:
            return

        application_info = await client.application_info()
        is_owner = (message.author.id == application_info.owner.id)

        logger.info(
            "id=%s user=%s userId=%s guild=%s guildId=%s | %s",
            message.id,
            message.author,
            message.author.id,
            message.guild.name,
            message.guild.id,
            to_one_liner(message.content)
        )

        request_body = {
            "id": f"discordpy:{message.id}",
            "message": message_text,
            "is_owner": is_owner,
            "character_limit": 2000,
            "format_name": "discord"
        }

        try:
            async with client_session.post(
                f"http://localhost:{backend_port_number}",
                json=request_body,
                timeout=10
            ) as response:
                reply_stack = await response.json()
            error = response.status != 200
            for reply in reply_stack:
                if error:
                    logger.error("id=%s | %s", message.id, to_one_liner(reply))
                else:
                    logger.info("id=%s | %s", message.id, to_one_liner(reply))
                await message.channel.send(reply)

        except Exception as error:
            logger.error("id=%s | %s", message.id, to_one_liner(str(error)))
            if isinstance(error, aiohttp.client_exceptions.ClientConnectorError):
                await message.channel.send("My brain stopped working. Please contact my owner. :<")
            else:
                await message.channel.send(str(error))

    client.run(config["discord_token"])


if __name__ == "__main__":
    main()
